compiler/rustc_mir/src/borrow_check/type_check/mod.rs

```diff
@@ -2051,7 +2048,14 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
                         // function definition. When we extract the
                         // signature, it comes from the `fn_sig` query,
                         // and hence may contain unnormalized results.
-                        let fn_sig = self.normalize(fn_sig, location);
+                        let mut fn_sig = self.normalize(fn_sig, location);
+
+                        if let ty::FnPtr(pol_sig) = ty.kind() {
+                            let mut sig = fn_sig.skip_binder();
+                            sig.constness = pol_sig.constness();
+                            sig.unsafety = pol_sig.unsafety();
+                            fn_sig = Binder::bind(sig);
+                        }
```

Comment:
Panicks without it (broken MIR in DefId(0:6 ~ rustconst[63c6]::main) (foo as fn() (Pointer(ReifyFnPointer))): equating const fn() with fn() yields NoSolution)

```
const fn foo() { }
const fn bar() { }
fn baz() { }


fn main() {
    let x: fn() = match 2 {
        2 => bar,
        3 => foo,
        _ => baz
    };

}
```
I think that it is not a better solution